rfcs/0009-vm-syscalls/0009-vm-syscalls.md

```diff
@@ -0,0 +1,234 @@
+
+```c
+static inline long
+__internal_syscall(long n, long _a0, long _a1, long _a2, long _a3, long _a4, long _a5)
+{
+  register long a0 asm("a0") = _a0;
+  register long a1 asm("a1") = _a1;
+  register long a2 asm("a2") = _a2;
+  register long a3 asm("a3") = _a3;
+  register long a4 asm("a4") = _a4;
+  register long a5 asm("a5") = _a5;
+
+  register long syscall_id asm("a7") = n;
+
+  asm volatile ("scall"
+		: "+r"(a0) : "r"(a1), "r"(a2), "r"(a3), "r"(a4), "r"(a5), "r"(syscall_id));
+
+  return a0;
+}
+
+#define syscall(n, a, b, c, d, e, f) \
+        __internal_syscall(n, (long)(a), (long)(b), (long)(c), (long)(d), (long)(e), (long)(f))
+```
+
+(NOTE: this is adapted from [riscv-newlib](https://github.com/riscv/riscv-newlib/blob/77e11e1800f57cac7f5468b2bd064100a44755d4/libgloss/riscv/internal_syscall.h#L25))
+
+Now we can trigger the same *Exit* syscall more easily in C code:
+
+```c
+syscall(93, 10, 0, 0, 0, 0, 0);
+```
+
+Note that even though *Exit* syscall only needs one argument, our C wrapper requires us to fill in all 6 arguments. We can initialize other unused arguments as all 0. Below we would illustrate each syscall with a C function signature to demonstrate each syscall's accepted arguments. Also for clarifying reason, all the code shown in this RFC is assumed to be written in pure C.
+
+### Exit
+
+As shown above, *Exit* syscall has a signature like following:
+
+```c
+void exit(int8_t code)
+{
+  syscall(93, code, 0, 0, 0, 0, 0);
+}
+```
+
+*Exit* syscall don't need a return value since CKB VM is not supposed to return from this function. Upon receiving this syscall, CKB VM would terminate execution with the specified return code. This is the only way of correctly exiting a contract in CKB VM.
+
+### Load Transaction
+
+*Load Transaction* syscall has a signature like following:
+
+```c
+int ckb_load_tx(void* addr, uint64_t* len, size_t offset)
+{
+  return syscall(2049, addr, len, offset, 0, 0, 0);
+}
+```
+
+The arguments used here are:
+
+* `addr`: a pointer to a buffer in VM memory space denoting where we would load the serialized transaction data.
+* `len`: a pointer to a 64-bit unsigned integer in VM memory space, when calling the syscall, this memory location should store the length of the buffer specified by `addr`, when returning from the syscall, CKB VM would fill in `len` with the actual length of the buffer. We would explain the exact logic below.
+* `offset`: an offset specifying from which offset we should start loading the serialized transaction data.
+
+When calling, this syscall would take the current transaction, and remove:
+
+* `unlock` scripts in all inputs
+* `data` part in all outputs
+* `contract` scripts in all outputs
+
+It then takes the modified transaction and serializes it into the FlatBuffers format. Then the serialized result is fed into VM via the steps below. For ease of reference, we refer the serialized result as `data`, and the length of `data` as `data_length`.
```

Comment:
should be CFB instead of FlatBuffer?

https://github.com/nervosnetwork/rfcs/pull/47